src/main/java/org/broadinstitute/hellbender/utils/runtime/StreamingProcessController.java

```diff
@@ -0,0 +1,598 @@
+     *
+     * @param settings                 Settings to be run.
+     * @param promptForSynchronization Prompt to be used as a synchronization point/boundary for blocking calls that
+     *                                 retrieve process output.
+     * @param enableJournaling         Turn on process I/O journaling. This records all inter-process communication to a file.
+     *                                 Journaling incurs a performance penalty, and should be used for debugging purposes only.
+     */
+    public StreamingProcessController(
+            final ProcessSettings settings,
+            final String promptForSynchronization,
+            final boolean enableJournaling) {
+        Utils.nonNull(settings, "Process settings are required");
+        this.settings = settings;
+        this.promptForSynchronization = promptForSynchronization;
+
+        if (enableJournaling) {
+            processJournal.enable(settings.getCommandString());
+        }
+    }
+
+    /**
+     * Starts the remote process running based on the setting specified in the constructor.
+     *
+     * @return true if the process has been successfully started
+     */
+    public boolean start() {
+        if (process != null) {
+            throw new IllegalStateException("This controller is already running a process");
+        }
+        process = launchProcess(settings);
+        startListeners();
+        processStdinStream = getProcess().getOutputStream();
+
+        return process.isAlive();
+    }
+
+    /**
+     * Write some input to the remote process, without waiting for output.
+     *
+     * @param line data to be written to the remote process
+     */
+    public void writeProcessInput(final String line) {
+        try {
+            // Its possible that we already have completed futures from output from a previous command that
+            // weren't consumed before we found the prompt at the last synchronization point. If so, drop them
+            // before we issue a new command (since retaining them could confound synchronization on output from
+            // this command), and warn.
+            if (stdErrFuture != null && stdErrFuture.isDone()) {
+                //logger.warn("Dropping stale stderr output: {}", stdErrFuture.get().getBufferString());
+                processJournal.writeLogMessage("Dropping stale stderr output: " + stdErrFuture.get().getBufferString());
+                stdErrFuture = null;
+            }
+            if (stdOutFuture != null && stdOutFuture.isDone()) {
+                //logger.warn("Dropping stale stdout output: {}", stdOutFuture.get().getBufferString());
+                processJournal.writeLogMessage("Dropping stale stdout output: " + stdOutFuture.get().getBufferString());
+                stdOutFuture = null;
+            }
+        } catch (InterruptedException e) {
+            throw new RuntimeException(String.format("Interrupted retrieving stale future: " + line, e));
```

Comment:
I'm curious why you're using RuntimeExceptions, it seems like either a GatkException or a specific ProcessControllerException would be better?